File: software/ai/sdk-go/policy/allowlist.go
package policy

import (
	"encoding/hex"

	"github.com/ethereum/go-ethereum/common"
	"github.com/ethereum/go-ethereum/crypto"
)

func NewContractAllowlist(contracts []common.Address) *ContractAllowlist {
	al := &ContractAllowlist{
		Contracts: make(map[common.Address]bool, len(contracts)),
	}
	for _, c := range contracts {
		al.Contracts[c] = true
	}
	return al
}

func NewFunctionAllowlist(signatures []string) *FunctionAllowlist {
	al := &FunctionAllowlist{
		Functions: make(map[string]bool, len(signatures)),
	}
	for _, sig := range signatures {
		selector := crypto.Keccak256([]byte(sig))[:4]
		al.Functions[hex.EncodeToString(selector)] = true
	}
	return al
}

func IsAllowed(p *Policy, contract common.Address, functionSig string) bool {
	if p == nil {
		return false
	}

	if p.ContractAllowlist != nil {
		if !p.ContractAllowlist.Contracts[contract] {
			return false
		}
	}

	if p.FunctionAllowlist != nil {
		selector := crypto.Keccak256([]byte(functionSig))[:4]
		selectorHex := hex.EncodeToString(selector)
		if !p.FunctionAllowlist.Functions[selectorHex] {
			return false
		}
	}

	return true
}

File: software/ai/sdk-ts/src/policy/allowlist.ts
import { type Address, type Hex, toFunctionSelector } from "viem";
import type { ContractAllowlist, FunctionAllowlist } from "../types/policy.js";
import { validateAddress } from "../utils/validation.js";

export function createContractAllowlist(addresses: Address[]): ContractAllowlist {
  if (addresses.length === 0) {
    throw new Error("Contract allowlist must contain at least one address");
  }

  const uniqueAddresses = [...new Set(addresses.map((a) => a.toLowerCase()))] as Address[];
  uniqueAddresses.forEach(validateAddress);

  return {
    type: "contract-allowlist",
    addresses: uniqueAddresses,
  };
}

export function createFunctionAllowlist(
  contract: Address,
  signatures: string[]
): FunctionAllowlist {
  validateAddress(contract);

  if (signatures.length === 0) {
    throw new Error("Function allowlist must contain at least one signature");
  }

  const selectors: Hex[] = signatures.map((sig) => {
    if (sig.startsWith("0x") && sig.length === 10) {
      return sig as Hex;
    }
    return toFunctionSelector(sig);
  });

  const uniqueSelectors = [...new Set(selectors)] as Hex[];

  return {
    type: "function-allowlist",
    contract,
    selectors: uniqueSelectors,
  };
}

export function mergeContractAllowlists(...allowlists: ContractAllowlist[]): ContractAllowlist {
  const allAddresses = allowlists.flatMap((a) => a.addresses);
  return createContractAllowlist(allAddresses);
}

export function mergeContractAndFunctionAllowlists(
  contracts: ContractAllowlist,
  functions: FunctionAllowlist[]
): { contracts: ContractAllowlist; functions: FunctionAllowlist[] } {
  const functionContracts = functions.map((f) => f.contract);
  const allContracts = [...contracts.addresses, ...functionContracts];
  return {
    contracts: createContractAllowlist(allContracts),
    functions,
  };
}
